File: src/while3addr/interpreter.ml
open Core
open Lang

let initial_env = String.Map.empty
let get_int a = match a with Int x -> x | _ -> failwith "Not an Intger"
let get_arr a = match a with Int b -> failwith "Not an array" | Arr c -> c
let handle_op op a b = Int (op (get_int a) (get_int b))
let get_val arr v = Array.get (get_arr arr) (get_int v)
let handle_bool op a b = op (get_int a) (get_int b : int)

let update_arr env arrName ind newval lookup =
  let arr = get_arr (lookup env arrName) in
  Array.set arr ind newval

let eval_insn env pc insn =
  let update env id value = String.Map.set env ~key:id ~data:value in
  let lookup = String.Map.find_exn in
  match insn with
  | ConstAssign (v, n) ->
      let new_env = update env v (Int n) in
      `Continue (pc + 1, new_env)
  | VarAssign (v1, v2) ->
      let n = lookup env v2 in
      let new_env = update env v1 n in
      `Continue (pc + 1, new_env)
  | OpAssign (v, v1, v2, op) ->
      let n1 = lookup env v1 in
      let n2 = lookup env v2 in
      let int_op =
        match op with
        | Add -> handle_op ( + )
        | Sub -> handle_op ( - )
        | Mul -> handle_op ( * )
        | Div -> handle_op ( / )
        | Get -> fun arr v -> Int (Array.get (get_arr arr) (get_int v))
      in
      let result = int_op n1 n2 in
      let new_env = update env v result in
      `Continue (pc + 1, new_env)
  | Goto location -> `Continue (location, env)
  | IfGoto (v, opr, lineno) ->
      let n = lookup env v in
      let comparison = match opr with LT -> ( < ) | EQ -> ( = ) in
      if comparison (get_int n) 0 then `Continue (lineno, env)
      else `Continue (pc + 1, env)
  | Print v ->
      let printer x =
        print_int x;
        print_string " "
      in
      let n = lookup env v in
      let () =
        match n with
        | Int a -> Format.printf "%d\n" (get_int n)
        | Arr b ->
            Array.iter b ~f:printer;
            print_string "\n"
      in
      `Continue (pc + 1, env)
  | Halt -> `Halt
  | ConstAssignArray (name, len) ->
      let arr = Array.create ~len 0 in
      let env' = String.Map.remove env name in
      let new_env = update env' name (Arr arr) in
      `Continue (pc + 1, new_env)
  | VarAssignArray (name, n) ->
      let len = get_int (lookup env name) in
      let arr = Array.create ~len 0 in
      let env' = String.Map.remove env name in
      let new_env = update env' name (Arr arr) in
      `Continue (pc + 1, new_env)
  | UpdateCC (arrName, ind, newVal) ->
      update_arr env arrName ind newVal lookup;
      `Continue (pc + 1, env)
  | UpdateII (arrName, x, y) ->
      let ind = get_int (lookup env x) in
      let newVal = get_int (lookup env y) in
      update_arr env arrName ind newVal lookup;
      `Continue (pc + 1, env)
  | UpdateIC (arrName, x, newVal) ->
      let ind = get_int (lookup env x) in
      update_arr env arrName ind newVal lookup;
      `Continue (pc + 1, env)
  | UpdateCI (arrName, ind, y) ->
      let newVal = get_int (lookup env y) in
      update_arr env arrName ind newVal lookup;
      `Continue (pc + 1, env)

let rec eval_program env (pc, listing) =
  let fetch location =
    match Int.Map.find listing location with
    | None -> Format.sprintf "No instruction at %d" location |> failwith
    | Some instruction -> instruction
  in
  match eval_insn env pc (fetch pc) with
  | `Continue (next_pc, new_env) -> eval_program new_env (next_pc, listing)
  | `Halt -> ()

File: src/while3addr/interpreter.py
from typing import *
from util import *
import sys
import concolic


ops = {
    '+': lambda x, y: x + y,
    '-': lambda x, y: x - y,
    '*': lambda x, y: x * y,
    '/': lambda x, y: x // y,
    '!!': lambda x, y: x[y]
}
cmps = {
    '<': lambda x, y: x < y,
    '=': lambda x, y: x == y
}
def getToken(instr):
    if instr[0] == 'halt':
        return 'halt'
    match [instr[0], instr[1]]:
        case [x, ":="]:
            if len(instr) == 3:
                v = instr[2]
                if v.isnumeric():
                    return 'constAssign'
                return 'varAssign'
            elif instr[2] == 'array':
                v = instr[3]
                if v.isnumeric():
                    return 'constArrayAssign'
                return 'varArrayAssign'
            elif instr[2] == 'input':
                return 'input'
            elif instr[2] == 'len':
                return 'len'
            return 'opAssign'
        case ['goto', _]:
            return 'goto'
        case ['if', _]:
            return 'if'
        case ['print', _]:
            return 'print'
        case ['update', _]:
            return 'update'
        case ['input', _]:
            return 'input'
        case other:
            print
            return 'fail'
    
# getProg("fun_progs/selection_sort.w3a", prog)
def eval_insn(env, args, pc : int, instr : List[str]) -> int:
    global ops, cmps
    def update(env, id, value):
        env[id] = value
    match getToken(instr):
        case 'input':
            # input file line var
            env[instr[0]] = args[int(instr[-1])]
            return pc + 1
        case 'len':
            arr = env[instr[-1]]
            env[instr[0]] = len(arr)
            return pc + 1
        case 'halt':
            return -1
        case 'constAssign':
            v = instr[2]
            update(env, instr[0], int(v))
            return pc + 1
        case 'varAssign':
            v = env[instr[2]]
            
            update(env, instr[0], v)
            return pc + 1
        case 'constArrayAssign':
            v = instr[3]
            update(env, instr[0], [0 for _ in range(int(v))])
            return pc + 1
        case 'varArrayAssign':
            v = env[instr[3]]
            update(env, instr[0], [0 for _ in range(v)])
            return pc + 1
        case 'goto':
            return int(instr[1])
        case 'if':
            lhs = instr[1]
            op = cmps[instr[2]]
            line = instr[5]
            v = env[lhs]
            if op(v, 0):
                return int(line)
            return pc + 1
        case 'print':
            print(env[instr[1]])
            return pc + 1
        case 'update':
            arrName = instr[1]
            index = env[instr[2]]
            val = env[instr[3]]
            arr = env[arrName]
            arr[index] = val
            return pc + 1
        case 'opAssign':
            [lhs, _, v0, op, v1] = instr
            v0 = env[v0]
            v1 = env[v1]
            update(env, lhs, ops[op](v0, v1))
            return pc + 1
        case 'fail':
            raise Exception("Instruction not implemented")

def fetch(listing, location):
    if location not in listing:
        raise Exception(f"No instruction at {location}")
    return listing[location]
def eval_program(env, args, pc, listing):
    while pc != -1:
        instr = fetch(listing, pc)
        pc = eval_insn(env, args, pc, instr)
        # print(instr, pc)
        if pc > max(listing):
            raise Exception(f"Illegal line number {pc}")
def eval_program_dev(env, args, pc, listing):
    while pc != -1:
        instr = fetch(listing, pc)
        if instr[0] == 'print':
            return env[instr[1]]
        pc = eval_insn(env, args, pc, instr)
        # print(instr, pc)
        if pc > max(listing):
            raise Exception(f"Illegal line number {pc}")
# eval_program(env, 1, prog)

def main(argv=None):
    env = {}
    args = sys.argv[1:]
    file = args[0]
    if argv == None:
        if len(args) - 1 == 1:
            args = [eval(args[1])]
        else:
            args = [[int(i) for i in args[1:]]]
        if args == []:
            print("No input file given")
            return 1

    prog = getProg(file)
    if argv:
        eval_program(env, argv, 1, prog)
    else:
        eval_program(env, args, 1, prog)

if __name__ == '__main__':
    main()
